clients/src/test/java/org/apache/kafka/clients/consumer/internals/IdempotentCloserTest.java

```diff
@@ -0,0 +1,206 @@
+        assertFalse(ic.isClosed());
+
+        // maybeThrowIllegalStateException doesn't throw anything since the closer is still in its "open" state.
+        assertDoesNotThrow(() -> ic.maybeThrowIllegalStateException(() -> "test"));
+
+        // Post-close, our call to maybeThrowIllegalStateException will, in fact, throw said exception.
+        ic.close();
+        assertTrue(ic.isClosed());
+        assertThrows(IllegalStateException.class, () -> ic.maybeThrowIllegalStateException(() -> "test"));
+    }
+
+    /**
+     * Tests that if the invoked onClose callback throws an exception, that:
+     *
+     * <ol>
+     *     <li>The exception does not prevent the {@link IdempotentCloser} from being updated to the closed state</li>
+     *     <li>The exception is bubbled up to the user</li>
+     * </ol>
+     */
+    @Test
+    public void testErrorsInOnCloseCallbacksAreNotSwallowed() {
+        IdempotentCloser ic = new IdempotentCloser();
+
+        // Verify initial invariants.
+        assertFalse(ic.isClosed());
+
+        // Upon close, our onClose callback will throw an error. First ensure that it is thrown at the user.
+        assertThrows(RuntimeException.class, () -> ic.close(CALLBACK_WITH_RUNTIME_EXCEPTION));
+
+        // Make sure the IdempotentCloser is still closed, though.
+        assertTrue(ic.isClosed());
+    }
+
+    /**
+     * Tests that if the invoked onSubsequentClose callback throws an exception, that it is thrown from
+     * {@link IdempotentCloser#close(Runnable, Runnable)} so the user can handle it.
+     */
+    @Test
+    public void testErrorsInOnPreviousCloseCallbacksAreNotSwallowed() {
+        IdempotentCloser ic = new IdempotentCloser();
+
+        // Verify initial invariants.
+        assertFalse(ic.isClosed());
+
+        // Perform the initial close. No errors here.
+        ic.close(CALLBACK_NO_OP);
+        assertTrue(ic.isClosed());
+
+        // Perform the subsequent close and verify that the exception is bubbled up to the user.
+        assertThrows(RuntimeException.class, () -> ic.close(CALLBACK_NO_OP, CALLBACK_WITH_RUNTIME_EXCEPTION));
+        assertTrue(ic.isClosed());
+    }
+
+    /**
+     * Tests that the {@link IdempotentCloser} implementation of {@link AutoCloseable#close()} works as expected.
+     */
+    @Test
+    public void testAutoCloseable() {
+        AtomicInteger onCloseCounter = new AtomicInteger();
+
+        try (IdempotentCloser ic = new IdempotentCloser() {
+            /**
+             * This provides us with the ability to track that the default {@link AutoCloseable#close()} method
+             * is called when the variable goes out of scope of the try-with-resources block.
+             */
+            @Override
+            public void close() {
+                close(onCloseCounter::getAndIncrement);
+            }
+        }) {
+            assertFalse(ic.isClosed());
+        }
+
+        assertEquals(1, onCloseCounter.get());
```

Comment:
Doesn't hurt, but I think this may be out of the test scope given that the try-with-resources is the one that ensures that the autoclosable is closed, and we're not testing that. The previous simplified version of the test that you had was useful and good enough I would say. It tested what we want to ensure (that our closer is autoclosable). The try on ln 179 would not compile I expect if the idempotent closer were not closable